src/lsqecc/simulation/logical_patch_state_simulation.py

```diff
@@ -263,3 +292,184 @@ def apply_logical_operation(self, logical_op: llops.LogicalLatticeOperation):
+            if isinstance(op, llops.AncillaQubitPatchInitialization):
+                add_initial_ancilla_state(op.qubit_uuid, op.qubit_state)
+            elif isinstance(op, llops.MagicStateRequest):
+                add_initial_ancilla_state(op.qubit_uuid, DefaultSymbolicStates.Magic)
+
+        all_init_states = [
+            get_init_state(quuid) for idx, quuid in self.mapper.enumerate_patches_by_index()
+        ]
+        return LazyTensorOp(all_init_states)
+
+    def apply_logical_operation(self, logical_op: llops.LogicalLatticeOperation):
+        """Update the logical state"""
+
+        if not logical_op.does_evaluate():
+            raise Exception(
+                "apply_logical_operation called with non evaluating operation :" + repr(logical_op)
+            )
+
+        if isinstance(logical_op, llops.SinglePatchMeasurement):
+            measure_idx = self.mapper.get_idx(logical_op.qubit_uuid)
+            operand_idx, idx_within_operand = self.logical_state.get_idxs_of_qubit(measure_idx)
+            operand = self.logical_state.ops[operand_idx]
+
+            local_observable = ConvertersToQiskit.pauli_op(logical_op.op)
+            observable_global_to_operand = circuit_apply_op_to_qubit(
+                qkop.I ^ operand.num_qubits, local_observable, idx_within_operand
+            )
+
+            distribution = ProjectiveMeasurement.pauli_product_measurement_distribution(
+                observable_global_to_operand, operand
+            )
+            outcome: ProjectiveMeasurement.BinaryMeasurementOutcome = proportional_choice(
+                distribution
+            )
+            self.logical_state.ops[operand_idx] = outcome.resulting_state
+            logical_op.set_outcome(outcome.corresponding_eigenvalue)
+
+            if idx_within_operand == operand.num_qubits - 1:
+                self.logical_state.separate_last_qubit_of_operand(operand_idx)
+                # TODO use .permute + reindexing to detach any qubit, not just the last
+        elif isinstance(logical_op, llops.LogicalPauli):
+            op_idx = self.mapper.get_idx(logical_op.qubit_uuid)
+            operand_idx, idx_within_operand = self.logical_state.get_idxs_of_qubit(op_idx)
+            operand = self.logical_state.ops[operand_idx]
+
+            symbolic_state = circuit_apply_op_to_qubit(
+                operand,
+                ConvertersToQiskit.pauli_op(logical_op.pauli_matrix),
+                idx_within_operand,
+            )
+            self.logical_state.ops[operand_idx] = cast(qkop.StateFn, symbolic_state.eval())
+
+        elif isinstance(logical_op, llops.MultiBodyMeasurement):
+            # Prepare the state by moving all involved operands to the front
+            self.align_state_with_pauli_op_to_first_operand(logical_op)
+            size_of_first_operand = self.logical_state.ops[0].num_qubits
+
+            # Make a global observable for the operator
+            global_observable = qkop.I ^ size_of_first_operand
```

Comment:
After discussion with @gwwatkin, for Lazy Tensor this means `first_operand_observable`, with the first_operand effectively causing the multibody measurement . `global_observable` might be misleading

Reply:
Applied suggestion